### PR TITLE
Fix #22

### LINKED ISSUE #22: Incorrect `Combustor` implementation
Let us go over one of the basic ideas of Pr47 asynchronous implementation: when an asynchronous FFI function wants to access the VM, it must acquire the running permission first, in order not corrupting other tasks. However, according to the current `Combustor` implementation:

https://github.com/Pr47/Pr47/blob/6d2d9ca5c58ec475746704ad01b2189aa5859ccb/pr47-core/src/vm/al31f/mod.rs#L55

This would allow one permission to be shared across several places, which is definitely inkorrect.

### PR DESCRIPTION
**Pull request by CousinZe**
This basically does #14, and fixes the problem presented in #22

### CODE REVIEW

File: pr47-core/src/vm/al31f/executor/async_executor.rs [hunk 1]
```diff
@@ -539,8 +541,8 @@ async unsafe fn vm_thread_run_function_impl<A: Alloc>(
                 }
                 ffi_args.set_len(args.len());
 
-                let mut combustor: Combustor<A> = Combustor::new(
-                    NonNull::from(&thread.vm)
+                let mut combustor: AsyncCombustor<A> = AsyncCombustor::new(
+                    thread.vm.derive_child_serializer_no_task()
```
Comment (CousinZe) [inline pr47-core/src/vm/al31f/executor/async_executor.rs]: This change is important. After doing so, permission won't be shared any more. However, by this time there's no way to "check" (either at compile time, or at run time) if a serializer has a permission, making our system hard to use and unsound.
@chuigda @SchrodingerZhu

Comment (chuigda) [reply pr47-core/src/vm/al31f/executor/async_executor.rs]: Nice, fine, give it one more!

File: pr47-core/Cargo.toml [hunk 2]
```diff
@@ -12,8 +12,7 @@ license = "MIT"
 pollster = "0.2"
 pr47-codegen = "0.0.2"
 unchecked_unwrap = "3"
-xjbutil = { version = "0.4", default-features = false, features = [
```
Comment (chuigda) [inline pr47-core/Cargo.toml]: This does not seem right to me.

**Review by chuigda — CHANGES_REQUESTED** [review]

**Comment by chuigda** [comment]
Try this.
<table style="border: 1px solid black">
    <tbody><tr><td>Before</td><td>After</td></tr>
    <tr>
      <td><code>type Serializer {
  shared: Mutex
  lock: ?MutexLock
}</code></td>
      <td><code>type Serializer {
  shared: Mutex
}

type LockedSerializer {
  lock: MutexLock
}</code></td>
    </tr>
  </tbody></table>

File: pr47-core/Cargo.toml (hunk 2 above)
Comment (CousinZe) [reply pr47-core/Cargo.toml]: Fixed.

File: pr47-core/src/ffi/async_fn.rs [hunk 3]
```diff
@@ -13,12 +13,12 @@ use crate::data::wrapper::{
...
-use crate::util::serializer::Serializer;
+use crate::util::serializer::{CoroutineSharedData, Serializer};
 
-pub trait AsyncVMContext: VMContext + Send + Sync {
-    type SharedData;
+pub trait AsyncVMContext: 'static + Sized + Send + Sync {
+    type VMData;
 
-    fn serializer(&self) -> &Serializer<Self::SharedData>;
+    fn serializer(&self) -> &Serializer<(CoroutineSharedData, Self::VMData)>;
```
Comment (CousinZe) [inline pr47-core/src/ffi/async_fn.rs:21]: Mark 1

File: pr47-core/src/util/serializer.rs [hunk 4]
```diff
@@ -21,11 +21,59 @@ use xjbutil::async_utils::{Mutex, MutexGuard, join_all, oneshot, task, yield_now
 use xjbutil::async_utils::oneshot::{Receiver, Sender};
 use xjbutil::unchecked::{UncheckedCellOps, UncheckedOption};
 
-/// Basic context shared by multiple `Serializer`s in the same *serialization group*.
-///
-/// The `SharedContext` serves as the manager of children tasks, task IDs and task completion
-/// signals. Read documentations of methods and fields for more information.
-pub struct SharedContext {
+/// A `Arc<Mutex>` is basically a "serializer" context, serializing accesses to `Data`.
+pub type Serializer<Data> = Arc<Mutex<Data>>;
```
Comment (CousinZe) [inline pr47-core/src/util/serializer.rs:25]: Mark 2

File: pr47-core/src/util/serializer.rs [hunk 5]
```diff
@@ -74,89 +122,119 @@ impl SharedContext {
...
-type Permit<SerializedData> = MutexGuard<'static, (SharedContext, SerializedData)>;
-
-/// Serializer context of one task
-pub struct Serializer<SerializedData: 'static> {
-    /// Context shared by all tasks in the same serialization group.
-    shared: Arc<Mutex<(SharedContext, SerializedData)>>,
-    /// Running permission of the current task.
-    permit: UnsafeCell<UncheckedOption<Permit<SerializedData>>>,
+/// Context of one coroutine/task
+pub struct CoroutineContext<SerializedData: 'static + Send> {
```
Comment (CousinZe) [inline pr47-core/src/util/serializer.rs:126]: Mark 3

File: pr47-core/src/vm/al31f/mod.rs [hunk 6]
```diff
@@ -51,39 +48,28 @@ impl<A: Alloc> VMContext for Combustor<A> {
 }
 
 #[cfg(feature = "async")]
-pub struct Combustor<A: Alloc> {
-    vm: NonNull<Serializer<AL31F<A>>>
+pub struct AsyncCombustor<A: Alloc> {
```
Comment (CousinZe) [inline pr47-core/src/vm/al31f/mod.rs:51]: Mark 4

**Review by CousinZe — COMMENTED** [review]
This change is important: we splitted up `Serializer` and `CoroutineContext`. The latter always comes with implicitly acquired running permission, while the former does not. Then, since `AsyncCombustor` now only contains a `Serializer` without running permission, it's now ok to safely use them in Async FFI functions.

File: pr47-core/src/util/serializer.rs (hunk 4 above)
Comment (chuigda) [reply pr47-core/src/util/serializer.rs:25]: Well done.

**Review by chuigda — APPROVED** [review]
Everything looks fine, except for the comments, which I could edit on myself.

**Review by SchrodingerZhu — APPROVED** [review]
LGTM

**Pull request merged**